New Uno Input, + Mega code organizing/extracting

Hyrum-Coleman opened the pull request.

Hyrum-Coleman committed:
reverse if condition to reduce nesting

Hyrum-Coleman committed:
no read from file, but did extract common logic to best of my ability

Hyrum-Coleman committed:
nullptr trolling

Hyrum-Coleman committed:
documentation

Hyrum-Coleman committed:
extract queue pop, remember to turn off motors

Hyrum-Coleman committed:
more extraction

Hyrum-Coleman committed:
change sending loop

sends predefined input on startup, sends input from Serial if provided in the loop function

nawper02 commented on Code/Robot/Robot/Robot.ino:
```diff
@@ -7,6 +7,8 @@
 #include <queue>
 #include "Wheelbase.h"
 
+const int NUMBER_OF_MOTORS = 4;
```
Rename to number of wheel motors? Considering we have 6 total motors

nawper02 commented on Code/Robot/Robot/Robot.ino:
```diff
@@ -52,63 +54,19 @@ void loop(JsonDocument doc) {
+        if (moveQueue->empty()) {
+          state = WAITING_TO_START;  // Go back to waiting state if queue is empty
         }
+        driving_logic(moveQueue);
```
Clean.

nawper02 commented on Code/Robot/Robot/Robot.ino:
```diff
@@ -126,3 +84,79 @@ void read_serial(JsonDocument& doc) {
+void driving_logic(queue<Move>* moveQueue) {
+  float motorSpeeds[NUMBER_OF_MOTORS]; // Initialize motor speeds
+
+  Move nextMove = getNextMoveFromQueue(moveQueue);
```
CLEAN

nawper02 commented on Code/Robot/Robot/Robot.ino:
```diff
@@ -126,3 +84,79 @@ void read_serial(JsonDocument& doc) {
+  return retMove;
+}
+
+void run_motors_with_blocking_delay(int delayTime, float* motorSpeeds, bool lift_motor) {
```
If the input to this was an int[6], we could set the corresponding motors (including lift motors) more cleanly. 

* with that said, I'm a big fan of this function

nawper02 commented on Code/Robot/Robot/Robot.ino:
```diff
@@ -126,3 +84,79 @@ void read_serial(JsonDocument& doc) {
+  }
+}
+
+void map_motor_speeds(float* motorSpeeds) {
```
Can we make this take in the max speed too? Might be more general if we can specify that when we call it instead of having the magic 200's in the map

nawper02 reviewed (changes requested):
Looks fantastic. A few ideas left in comments. Would be happy merging this as is tho

Hyrum-Coleman committed:
include max speed in map function call

Hyrum-Coleman committed:
another extraction

Hyrum-Coleman committed:
rename const

Hyrum-Coleman committed:
remove using namespace

we weren't using it and it's bringing in extra stuff we don't want necessarily

Hyrum-Coleman committed:
extra whitespace removal

Hyrum-Coleman committed:
change symbols from motor to wheel where appropriate

Hyrum-Coleman committed:
rename time, move maxSpeed to variable for easier changes

Hyrum-Coleman committed:
fix the consequences of removing 'using namespace std'

Hyrum-Coleman committed:
change input json to be current copy paste for all motors 1 sec

nawper02 committed:
Add few lines to UnoSending so we can see messages coming from mega

nawper02 committed:
Fix memory leak

std::queue<Move>* moveQueue = new std::queue<Move>();
 inside loop?!?!?!?!?!?!?!?!?!? What da

nawper02 committed:
Nevermind about the memory leak

nawper02 committed:
ITS ALIVE!

Fixed repeating action bug (was because we popped the front but took the back of the movequeue)

Fixed not going back to start state bug (added 'break')

nawper02 committed:
Add debug print macro

nawper02 committed:
Functioning. Need to fix flipped left and right

nawper02 committed:
PM6 Code.

nawper02 committed:
L298N not working

Added custom barebones motor driver class. that doesnt work either.

nawper02 committed:
Add custom barebones library

nawper02 committed:
This version is working with both libraries (barebones and theirs). But, It is still not driving the L298N motors backwards.

Current issues:
1. only drives one of two l298n motors
2. once that motor has been driven once, it cannot be driven again for the lifetime of the program.
3. Cannot reverse the l298n motor, probably because of problem 2.

Hyrum-Coleman committed:
Change names to match C/C++ naming conventions better

it's for the best

Hyrum-Coleman committed:
const function in wheelbase :bangbang:

Makes the wheelbase do it's computation at comptime. Promises the functions won't change the member variables attached to the wheelbase (We shouldn't need to).

Dremich committed:
motorDebugging for PM6?

Dremich committed:
Added Uno wireless com code

nawper02 committed:
Robot.ino should be back to int main(). Hyrum PLEASE check to make sure I did this right

nawper02 committed:
Add comments  -- troubleshooting suggestions

Hyrum-Coleman committed:
PM6 passing (on god)

nawper02 committed:
Add 99 bogus pins to diagnostic and current monitoring

nawper02 committed:
Merge branch 'read_from_file' of https://github.com/Hyrum-Coleman/mechatronics-goats into read_from_file

Hyrum-Coleman committed:
make conflicting pins go away

Hyrum-Coleman committed:
Merge branch 'read_from_file' of https://github.com/Hyrum-Coleman/mechatronics-goats into read_from_file

nawper02 committed:
Update pintable -- A9 for CM conflicts

nawper02 committed:
Add library for reflectance array

The github has SO MUCH useful info

nawper02 committed:
Add SensorBumper.SLDPRT

nawper02 reviewed (approved):
Can we merge?
